server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,242 @@
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+
+    @Nullable
+    @Override
+    public TopKState newState(RamAccounting ramAccounting,
+                              Version indexVersionCreated,
+                              Version minNodeInCluster,
+                              MemoryManager memoryManager) {
+        ramAccounting.addBytes(calculateRamUsage(DEFAULT_MAP_SIZE));
+        return new TopKState(new ItemsSketch<>(DEFAULT_MAP_SIZE), DEFAULT_LIMIT);
+    }
+
+    @Override
+    public TopKState iterate(RamAccounting ramAccounting,
+                             MemoryManager memoryManager,
+                             TopKState state,
+                             Input<?>... args) throws CircuitBreakingException {
+        if (state.itemsSketch.isEmpty() && args.length == 2) {
+            Integer limit = (Integer) args[1].value();
+            if (limit <= 0) {
+                throw new IllegalArgumentException("Invalid value for parameter k for top-k aggregation");
+            }
+            int maxMapSize = maxMapSize(limit);
+            // Calculate new max space usage and adapt ram accounting
+            long newRamUsage = calculateRamUsage(maxMapSize);
+            long defaultRamUsage = calculateRamUsage(DEFAULT_MAP_SIZE);
+            long diff = newRamUsage > defaultRamUsage ? newRamUsage - defaultRamUsage : defaultRamUsage - newRamUsage;
+            ramAccounting.addBytes(diff);
+            state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+        }
+        Object value = args[0].value();
+        state.itemsSketch.update(value);
+        return state;
+    }
+
+    private int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        return maxMapSize * 18L;
+    }
+
+    @Override
+    public TopKState reduce(RamAccounting ramAccounting, TopKState state1, TopKState state2) {
+        return new TopKState(state1.itemsSketch.merge(state2.itemsSketch), Math.min(state1.limit, state2.limit));
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, TopKState state) {
+        ItemsSketch.Row<Object>[] frequentItems = state.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+        int limit = Math.min(frequentItems.length, state.limit);
+        var result = new ArrayList<>(limit);
+        for (int i = 0; i < limit; i++) {
+            var item = frequentItems[i];
+            var row = new HashMap<>(2);
```

Comment:
```suggestion
            Map<String, Object> row = HashMap.newHashMap(2);
```

Otherwise it is typed as `Map<Object, Object>` and the sizing doesn't take the load factor into consideration.

But even better would probably be:

```
            var item = frequentItems[i];
            result.add(Map.of(
                "item", item.getItem(),
                "frequency", item.getEstimate()
            ));
```

Because `Map.of` can use a more efficient structure (key and values are interleaved on the same array)